src/game/UIExitMenu.cpp

```diff
@@ -0,0 +1,180 @@
+// UIChat.cpp: implementation of the CUIChat class.
+//
+//////////////////////////////////////////////////////////////////////
+
+#include "StdAfx.h"
+#include "UIExitMenu.h"
+#include "PlayerMySelf.h"
+
+#include "APISocket.h"
+#include "PacketDef.h"
+#include "GameProcMain.h"
+#include "GameProcedure.h"
+#include "GameProcLogin.h"
+#include "GameDef.h"
+
+#include "UIMessageWnd.h"
+#include "UIManager.h"
+#include "UIPartyOrForce.h"
+#include "SubProcPerTrade.h"
+
+#include "Resource.h"
+
+//////////////////////////////////////////////////////////////////////
+// Construction/Destruction
+//////////////////////////////////////////////////////////////////////
+
+CUIExitMenu::CUIExitMenu() {
+    m_pBtn_chr = NULL;
+    m_pBtn_exit = NULL;
+    m_pBtn_option = NULL;
+    m_pBtn_cancel = NULL;
+}
+
+CUIExitMenu::~CUIExitMenu() {}
+
+void CUIExitMenu::Release() {
+    CN3UIBase::Release();
+}
+
+void CUIExitMenu::SelectCharacter() {
+
+    /*
+	
+  v0 = s_pPlayer;
+    if (s_pPlayer && s_pFX) {
+        s_pPlayer->m_bMoveUnk = 0;
+        s_pPlayer->field_9EC = 0;
+        CN3FXMgr::TriggerCursor((_DWORD *)s_pFX, s_pPlayer->m_InfoBase_iID, s_pPlayer->m_InfoBase_iID,
+                                30001, // fx/target_pointer.fxb
+                                30001, FX_BUNDLE_MOVE_DIR_FLEXABLETARGET);
+        v0 = s_pPlayer;
+    }
+    
+    */
+    // TODO: CN3FXMgr::TriggerCursor
+    CGameBase::s_pPlayer->StunRelease();
+
+    if (CGameProcedure::s_pProcMain->m_pSubProcPerTrade) {
+
+        // TODO: MERCHANT CLEAR MODEL
+
+        CGameProcedure::s_pProcMain->m_pSubProcPerTrade->LeavePerTradeState(PER_TRADE_RESULT_MY_CANCEL);
+    }
+
+    if (CGameProcedure::s_pProcMain->m_pUIPartyOrForce) {
+        CGameProcedure::s_pProcMain->m_pUIPartyOrForce->MemberDestroy();
+    }
+
+    std::string szIP = CGameProcedure::s_pSocket->GetCurrentIP();
+    DWORD       dwPort = CGameProcedure::s_pSocket->GetCurrentPort();
+
+    CGameProcedure::s_bNeedReportConnectionClosed = false;
+    CGameProcedure::s_pSocket->Disconnect();
+    Sleep(4000);
+    CGameProcedure::s_pSocket->Connect(CGameProcedure::s_hWndBase, szIP.c_str(), dwPort);
+    CGameProcedure::s_bNeedReportConnectionClosed = true;
+
+    CGameProcedure::MsgSend_GameServerLogIn();
+    CGameProcedure::ProcActiveSet((CGameProcedure *)CGameProcedure::s_pProcCharacterSelect);
+}
+
+void CUIExitMenu::AddWarningMessage(DWORD idMessage, bool visbleMenu) {
+
+    if (!CGameProcedure::s_pProcMain->m_pUIChatDlg) {
+        return;
+    }
+
+    std::string szMsg;
+    ::_LoadStringFromResource(idMessage, szMsg);
+
+    CGameProcedure::s_pProcMain->m_pUIChatDlg->AddChatMsg(N3_CHAT_NORMAL, szMsg, 0xFFFF0000);
+
+    SetVisible(visbleMenu);
+}
+
+bool CUIExitMenu::ReceiveMessage(CN3UIBase * pSender, DWORD dwMsg) {
+    if (NULL == pSender) {
+        return false;
+    }
+
+    if (dwMsg == UIMSG_BUTTON_CLICK) {
+
+        if (pSender->m_szID == "btn_chr") { // Select characeter
```

Comment:
Let's not use id's to compare senders:
```suggestion
        if (pSender == (CN3UIBase *)m_pBtnChr) { // Select characeter
```

Reply:
Fixed here: https://github.com/ko4life-net/ko/pull/166/commits/dbbae38db8feb6e66b84f92f2cc381937d4cc785